feat(docs):  upgrade presentation of display.rst

vwheeler63 opened the pull request:
...based on email discussion with Gábor.

Display now has its own (more readable and thus more useful) TOC tree, [demonstrated here](http://crystal-clear-research.com/for_gabor/demo/docs.lvgl.io/master/details/main-components/index.html).

cc:  @kisvegabor 

### Notes
- Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed.  Done.
- Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant.  n/a
- Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.  n/a
- If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/master/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/master/Kconfig).  n/a
- Run `scripts/code-format.py` (`astyle v3.4.12` needs to installed by running `cd scripts; ./install_astyle.sh`) and follow the [Code Conventions](https://docs.lvgl.io/master/CODING_STYLE.html).  n/a
- Mark the Pull request as [Draft](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request) while you are working on the first version, and mark is as _Ready_ when it's ready for review.  Done.
- When changes were requested, [re-request review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) to notify the maintainers.  Done.
- Help us to review this Pull Request! Anyone can [approve or request changes](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews).

vwheeler63 commented:
not stale

vwheeler63 commented:
> * Display Features might not be required. Attributes and Screen Layers can have their own section directly.

Done.  Good observation!



> * The word "Display" is redundant in "Display Events" and "Display setup". As we are already in the "Display" section maybe "Events" and "Setup" would be also enough.

That's funny, in reviewing that, I saw the same redundancy for the same reason.  Handled!

> * Move each advanced features into their own file. Some of them are quite long, but the short ones can be completed a little too

Interesting thought.  Most of them are completely separate topics except 2 on refreshing which I kept together.  Additionally I found and handled these situations that I found:

- one event name changed:  LV_EVENT_FLUSH_READY => LV_EVENT_FLUSH_FINISH
- clarification of meanings of 3 events
- documentation of display events carried into `lv_event.h` where they were documentation was previously missing
- descriptions were missing for display events in `event.rst`
- LV_EVENT_FLUSH_WAIT_START and LV_EVENT_FLUSH_WAIT_FINISH were missing in documentation of display events
- minor unnecessary CPU overhead in `lv_refr.c` eliminated
- observed that display events are included in 2 places in the documentation; factored them into 1 place (`display_events_list.txt`) and did an `.. include::` in both places.
- added additional use case to `Decoupling the Display Refresh Timer` section (one I personally use to save CPU overhead)
- corrected name of `lv_display_refr_timer()` function in same section (seems it was a hold-over from v8)
- minor editing to improve wording in `Forcing a Refresh` section
- added reference to example code in `User Data` section
- minor editing for clarifications in `Flush-Wait Callback` section

These are all now submitted herein.

Unfortunately, breaking up the `display_advanced.rst` into individual files, but not doing so for the basics sections causes an inconsistency in the TOC.  It's not bad though, but it makes "basics" redundant and makes "Advanced Features" in its own font size.  However, I tried various other options manipulating the titling (e.g. getting the TOC to display "Basics" as a caption like "Advanced Features" font (in white), and remove the `display_basics.rst` title, but then then the "breadcrumb" list at the top of the `display_basics.rst` doc doesn't come out right -- it chooses the "What is a Display?" as the name of the last breadcrumb, which isn't appropriate.  And I do not thin that breaking up `display_basics.rst` is a good idea for this reason:  a beginning user is going to need all of those topics, which connects them.  True:  breaking it up would remedy the TOC inconsistency, but then sort of creates a doc maintenance problem because reading the documentation "in sequence" is now difficult with a mere text editor -- the docs have to be generated and read through a browser to see if the logic holds together.  :-(

![image](https://github.com/user-attachments/assets/6eead83a-b185-4570-b202-5e02c2ef8d1e)

So I am going to submit it as it is, and here is the list of files, but I am currently hesitant about the results.  This is the current file list:

![image](https://github.com/user-attachments/assets/b92dfc9e-104a-45c7-ba79-294c4b8ae70f)

The reason I am hesitant is:  to be consistent, we either need to:

- break up both (`display_basics.rst` as well as `display_advanced.rst`) -- which gives us a document maintenance problem -- now a text editor cannot be used to read through the display documentation to see if the logic sequence is good -- it has to be done by generating and reading it through a browser; or

- leave both `display_basics.rst` as well as `display_advanced.rst` intact, leaving the pages longer, but they are easier to maintain.

There is one other danger:  each of these `.html` files is about 750K, even if it only has 1 paragraph in it.  So anyone with slow Internet is going to feel it in page-loading times.  I don't know if that is a problem for anyone any more, or if it impacts phone accounts with limited total data before additional billing happens.  (I personally have chosen "unlimited" for my phone plan for the last 10 years, but I had to ask for it.  I'm not sure what the normal phone plan is like, and whether they still have data limits.)  Just a thought.

This link now gives what it currently looks like.  It also does this to the TOC above it, which isn't bad.  Thoughts?

![image](https://github.com/user-attachments/assets/d39f83d1-646a-465f-b7b6-0a63d29e9f24)

vwheeler63 commented:
I would like to mark this PR as "pending discussion" and not "submitted for approval" because it is in an inconsistent state as you can see from the discussion above.  I think more changes are needed to bring it back to a consistent state -- either breaking up `display_basics.rst` (introduces a doc-maintenance problem) or putting `display_advanced.rst` back together again.

kisvegabor commented:
Thank you Victor!

Just  a few notes:
What if we make the first pages like this:
   - Overview (I think it's a good first page in general)
   - Creating displays (the current setup section + draw buffers + Flush + Flush-Wait Callback)
   - Screen layers
   - Events
   - Resolution
   - Rotation
   - ...

This way  there is no Basic and advanced sections as it's obvious that "Overview" and "Creating displays" are the most important ones. In general we can order the pages to start with the most generic, and finish with the most specific/special ones.  
   
"Swapping Endian-ness", "Monochrome support" and "[Transparent screens](http://crystal-clear-research.com/for_gabor/demo/docs.lvgl.io/master/details/base-widget/obj.html#transparent-screens)" could be part of "Color format".

"Refreshing" looks like this [here](http://crystal-clear-research.com/for_gabor/demo/docs.lvgl.io/master/details/main-components/display.html)
![image](https://github.com/user-attachments/assets/d7c67321-243e-48b4-a39f-1e43c528e826) 
but like this [here](http://crystal-clear-research.com/for_gabor/demo/docs.lvgl.io/master/details/main-components/display_tiling.html)
![image](https://github.com/user-attachments/assets/be428c13-05f1-4d9b-b269-027df46ac12d)


"Mirroring a Display" and "Split Image" also can be on a joint page. "Extending/Combining display"?

vwheeler63 commented:
Hi, Gábor!

I like the way you think, sir!  :-)  The most important thing is that:

- it be in logical sequence so readers can find what they are looking for quickly while skipping things they already know, and
- later concepts build on more basic ones,

and I think you nailed it.  :-)  I will work on this now....

kisvegabor commented:
Cool! Looking forward to seeing the updated version!

vwheeler63 commented:
@liamHowatt This one is MOSTLY a reorganization of the table of contents, including sequence of presentation.  There are a number of small edits, but nothing major.

vwheeler63 commented:
@kisvegabor I made some slight alterations to the sequence of presentation, which I think preserves the concept of more-basic-and-more-frequently-used to more-detailed-and-less-frequently-used.  Let me know if (with your 40,000 overview, since you wrote most of the code) you find the topics could be in a better sequence.  It made sense to me based on my own (limited) experience using Display (`lv_display_t`) in firmware.

kisvegabor commented:
I'm checking the updates [here](http://crystal-clear-research.com/for_gabor/demo/docs.lvgl.io/master/details/main-components/display_color_format.html). It's updated right?

---

I've noticed that "Refreshing" changes its place. 
In "Displays"
![image](https://github.com/user-attachments/assets/ab790cd3-28f2-49dc-80ed-f7becb1ba18b)

In "Color formats"
![image](https://github.com/user-attachments/assets/f4047c9b-5a93-4018-9e93-a80eb6f1dfa6)

---

What's your opinion about these? 
> "Mirroring a Display" and "Split Image" also can be on a joint page. "Extending/Combining display"?
"Swapping Endian-ness", "Monochrome support" and "Transparent screens" could be part of "Color format".

---

What about creating a `display` folder instead having a lot of `display_....rst` files?
![image](https://github.com/user-attachments/assets/d5437dfe-db94-4592-a0f2-d06114439bf0)

vwheeler63 commented:
> I've noticed that "Refreshing" changes its place.

My fault:  I moved it in the TOC and then did not re-upload all of them.

> What's your opinion about these?
> 
> > "Mirroring a Display" and "Split Image" also can be on a joint page. "Extending/Combining display"?
> > "Swapping Endian-ness", "Monochrome support" and "Transparent screens" could be part of "Color format".

Brilliant!  To accommodate this structure change, I added a link to Transparent Screens where it used to be (in Base Widget > Widget Basics > Layers and added another one as "Further Reading" at the end of "Display > Screen Layers".  I also re-uploaded the whole doc set so everything (including the TOCs) are in sync.

> What about creating a `display` folder instead having a lot of `display_....rst` files?

Agreed!  Done!

kisvegabor reviewed (approved):
I love it! How far are we from finishing the proofreading?

vwheeler63 commented:
> I love it! How far are we from finishing the proofreading?

Other-Components > XML, Libs, and Integration remaining.  I'm hoping it starts getting faster.

vwheeler63 commented:
@liamHowatt Good morning, Liam!  FYI, this one is "low-hanging fruit" because there has been no significant re-writing (a few edits), but mostly this reorganizes the table of contents and placement, and 99.5% of the content was preserved as-is (adjusting the section-heading underlining to conform with its new "standing" in a separate document.  What was once:

```
details/main-components/
    display.rst
```

is now

```
details/main-components/display/
    index.rst
        +-- overview.rst
        +-- setup.rst
        +-- screen_layers.rst
        +-- color_format.rst
        +-- refreshing.rst
        +-- display_events.rst
        +-- resolution.rst
        +-- inactivity.rst
        +-- redraw_area.rst
        +-- tiling.rst
        +-- extending_combining.rst

where 1, 2 or 3 major sections out of the original `display.rst` were "factored out" into their own pages.

Commit 809a5e1 is the one with the relatively-minor content edits.

kisvegabor commented:
I added @liamHowatt as a reviewer.

kisvegabor commented:
@liamHowatt @uLipe please review this PR.

uLipe reviewed (approved):
Great work!

liamHowatt commented on docs/details/main-components/display/refreshing.rst:
```diff
@@ -0,0 +1,68 @@
+
+Normally the dirty (a.k.a invalid) areas are checked and redrawn in
+every :c:macro:`LV_DEF_REFR_PERIOD` milliseconds (set in ``lv_conf.h``).
+This happens as a result of a refresh :ref:`timer` created that gets created when
+the display is created
+executed
+at that
+interval
```
```suggestion
This happens as a result of a refresh :ref:`timer` that gets created when
the display is created and
executes
at that
interval.
```

liamHowatt commented on docs/details/main-components/display/refreshing.rst:
```diff
@@ -0,0 +1,68 @@
+   lv_display_refr_timer(NULL);
+
+If you have multiple displays call :cpp:expr:`lv_display_set_default(display1)` to
+select the display to refresh before :cpp:expr:`_lv_display_refr_timer(NULL)`.
+
+
+.. note:: :cpp:func:`lv_timer_handler` and :cpp:func:`_lv_display_refr_timer` must not run at the same time.
```
```suggestion
select the display to refresh before :cpp:expr:`lv_display_refr_timer(NULL)`.


.. note:: :cpp:func:`lv_timer_handler` and :cpp:func:`lv_display_refr_timer` must not run at the same time.
```

vwheeler63 replied:
Hi, @liamHowatt !  Wow, I did not see your suggestions before you committed this PR, but I like them, so I am adding them here locally to go into a subsequent PR.  Thank you!

vwheeler63 replied:
P.S.  Good catch with the `lv_display_refr_timer()` function names.  That must have changed recently, as I see it in the lv_api_map_v9_1.h file.  Again, thank you!

liamHowatt reviewed (commented):
Looks good! Thank you!
